File: src/features/workshop/api/workshop.ts
// src/features/workshop/api/workshop.ts
import { invoke } from "@/lib/tauri";
import type {
    AddLaborInput,
    AddPartInput,
    CreateCustomerVehicleInput,
    CreateServiceOrderInput,
    CustomerVehicle,
    ServiceOrder,
    ServiceOrderStatus,
} from "../types";

export function findCustomerVehicleByVin(vin: string): Promise<CustomerVehicle> {
    return invoke<CustomerVehicle>("find_customer_vehicle_by_vin", { vin });
}

export function listCustomerVehicles(customerId: string): Promise<CustomerVehicle[]> {
    return invoke<CustomerVehicle[]>("list_customer_vehicles", { customerId });
}

export function createCustomerVehicle(
    input: CreateCustomerVehicleInput
): Promise<CustomerVehicle> {
    return invoke<CustomerVehicle>("create_customer_vehicle", { input });
}

/** El número de orden (OS-...) lo genera el backend. */
export function createServiceOrder(input: CreateServiceOrderInput): Promise<ServiceOrder> {
    return invoke<ServiceOrder>("create_service_order", { input });
}

export function getServiceOrder(id: string): Promise<ServiceOrder> {
    return invoke<ServiceOrder>("get_service_order", { id });
}

export function listServiceOrdersByStatus(status: ServiceOrderStatus): Promise<ServiceOrder[]> {
    return invoke<ServiceOrder[]>("list_service_orders_by_status", { status });
}

export function updateServiceOrderStatus(
    id: string,
    newStatus: ServiceOrderStatus,
    diagnosis?: string | null
): Promise<ServiceOrder> {
    return invoke<ServiceOrder>("update_service_order_status", { id, newStatus, diagnosis });
}

export function addLabor(input: AddLaborInput): Promise<void> {
    return invoke<void>("add_labor", { input });
}

export function addPart(input: AddPartInput): Promise<void> {
    return invoke<void>("add_part", { input });
}

File: src-tauri/src/models/workshop.rs
// src-tauri/src/models/workshop.rs

use chrono::{DateTime, Utc};
use rust_decimal::Decimal;
use serde::{Deserialize, Serialize};
use uuid::Uuid;

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize, sqlx::Type)]
#[sqlx(
    type_name = "workshop.service_order_status",
    rename_all = "SCREAMING_SNAKE_CASE"
)]
#[serde(rename_all = "SCREAMING_SNAKE_CASE")]
pub enum ServiceOrderStatus {
    Recibido,
    Diagnostico,
    EnReparacion,
    EsperaRepuestos,
    Listo,
    Entregado,
    Cancelado,
}

#[derive(Debug, Clone, Serialize, Deserialize, sqlx::FromRow)]
#[serde(rename_all = "camelCase")]
pub struct CustomerVehicle {
    pub id: Uuid,
    pub customer_id: Uuid,
    pub vehicle_unit_id: Option<Uuid>,
    pub vin_chassis_number: Option<String>,
    pub engine_number: Option<String>,
    pub brand_id: Option<Uuid>,
    pub model_name: Option<String>,
    pub model_year: Option<i16>,
    pub plate_number: Option<String>,
    pub created_at: DateTime<Utc>,
}

#[derive(Debug, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct CreateCustomerVehicleInput {
    pub customer_id: Uuid,
    pub vehicle_unit_id: Option<Uuid>,
    pub vin_chassis_number: Option<String>,
    pub engine_number: Option<String>,
    pub brand_id: Option<Uuid>,
    pub model_name: Option<String>,
    pub model_year: Option<i16>,
    pub plate_number: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize, sqlx::FromRow)]
#[serde(rename_all = "camelCase")]
pub struct LaborCatalog {
    pub id: Uuid,
    pub name: String,
    pub description: Option<String>,
    pub standard_price: Decimal,
    pub estimated_hours: Option<Decimal>,
    pub is_active: bool,
}

#[derive(Debug, Clone, Serialize, Deserialize, sqlx::FromRow)]
#[serde(rename_all = "camelCase")]
pub struct ServiceOrder {
    pub id: Uuid,
    pub order_number: String,
    pub customer_vehicle_id: Uuid,
    pub warehouse_id: Uuid,
    pub status: ServiceOrderStatus,
    pub reported_issue: String,
    pub diagnosis: Option<String>,
    pub assigned_technician_id: Option<Uuid>,
    pub mileage_km: Option<i32>,
    pub labor_total: Decimal,
    pub parts_total: Decimal,
    pub total_amount: Decimal,
    pub received_at: DateTime<Utc>,
    pub promised_at: Option<DateTime<Utc>>,
    pub delivered_at: Option<DateTime<Utc>>,
    pub created_at: DateTime<Utc>,
    pub updated_at: DateTime<Utc>,
}

#[derive(Debug, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct CreateServiceOrderInput {
    pub customer_vehicle_id: Uuid,
    pub warehouse_id: Uuid,
    pub reported_issue: String,
    pub assigned_technician_id: Option<Uuid>,
    pub mileage_km: Option<i32>,
    pub promised_at: Option<DateTime<Utc>>,
}

#[derive(Debug, Clone, Serialize, Deserialize, sqlx::FromRow)]
#[serde(rename_all = "camelCase")]
pub struct ServiceOrderLabor {
    pub id: Uuid,
    pub service_order_id: Uuid,
    pub labor_id: Uuid,
    pub price_charged: Decimal,
    pub performed_by: Option<Uuid>,
}

#[derive(Debug, Clone, Serialize, Deserialize, sqlx::FromRow)]
#[serde(rename_all = "camelCase")]
pub struct ServiceOrderPart {
    pub id: Uuid,
    pub service_order_id: Uuid,
    pub product_id: Uuid,
    pub quantity: Decimal,
    pub unit_price: Decimal,
}

/// Input para agregar mano de obra a una orden ya creada.
#[derive(Debug, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct AddLaborInput {
    pub service_order_id: Uuid,
    pub labor_id: Uuid,
    pub price_charged: Decimal,
    pub performed_by: Option<Uuid>,
}

/// Input para consumir un repuesto en una orden — esto SÍ descuenta
/// stock real (a diferencia de agregar mano de obra, que no toca
/// inventario). Por eso viaja junto con lo necesario para el kardex.
#[derive(Debug, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct AddPartInput {
    pub service_order_id: Uuid,
    pub product_id: Uuid,
    pub quantity: Decimal,
    pub unit_price: Decimal,
}
